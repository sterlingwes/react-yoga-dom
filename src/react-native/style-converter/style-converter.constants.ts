/**
 * Copyright (c) Nicolas Gallagher.
 * (Adapted from react-native-web)
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

export const STYLE_SHORT_FORM_EXPANSIONS = {
  borderColor: ['borderTopColor', 'borderRightColor', 'borderBottomColor', 'borderLeftColor'],
  borderRadius: [
    'borderTopLeftRadius',
    'borderTopRightRadius',
    'borderBottomRightRadius',
    'borderBottomLeftRadius',
  ],
  borderStyle: ['borderTopStyle', 'borderRightStyle', 'borderBottomStyle', 'borderLeftStyle'],
  borderWidth: ['borderTopWidth', 'borderRightWidth', 'borderBottomWidth', 'borderLeftWidth'],
  margin: ['marginTop', 'marginRight', 'marginBottom', 'marginLeft'],
  marginHorizontal: ['marginRight', 'marginLeft'],
  marginVertical: ['marginTop', 'marginBottom'],
  overflow: ['overflowX', 'overflowY'],
  overscrollBehavior: ['overscrollBehaviorX', 'overscrollBehaviorY'],
  padding: ['paddingTop', 'paddingRight', 'paddingBottom', 'paddingLeft'],
  paddingHorizontal: ['paddingRight', 'paddingLeft'],
  paddingVertical: ['paddingTop', 'paddingBottom'],
};

export const MONOSPACE_FONT_STACK = 'monospace,monospace';
export const SYSTEM_FONT_STACK =
  'system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Ubuntu,"Helvetica Neue",sans-serif';
