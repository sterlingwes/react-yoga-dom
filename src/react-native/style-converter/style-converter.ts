/**
 * Copyright (c) Nicolas Gallagher.
 * (Adapted from react-native-web)
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import decamelize from 'decamelize';

import {
  MONOSPACE_FONT_STACK,
  SYSTEM_FONT_STACK,
  STYLE_SHORT_FORM_EXPANSIONS,
} from './style-converter.constants';
import { normalizeValueWithProperty } from './normalizer';

const toCSSProp = (prop: string) => decamelize(prop, '-');

/**
 * The browser implements the CSS cascade, where the order of properties is a
 * factor in determining which styles to paint. React Native is different. It
 * gives giving precedence to the more specific style property. For example,
 * the value of `paddingTop` takes precedence over that of `padding`.
 *
 * This module creates mutally exclusive style declarations by expanding all of
 * React Native's supported shortform properties (e.g. `padding`) to their
 * longfrom equivalents.
 */

type TransformInstruction = {
  [attr: string]: string | number;
};

type WidthHeight = {
  width: number;
  height: number;
};

type RNStyleObj = {
  // subset of RN styles for typechecking attributes we handle explicitly
  borderWidth?: number;
  borderWidthLeft?: number;
  borderWidthRight?: number;
  boxShadow?: string;
  direction?: string;
  display?: string;
  flex?: number;
  flexBasis?: string | number;
  flexShrink?: number;
  font?: string;
  fontFamily?: string;
  fontVariant?: string | Array<string>;
  marginVertical?: string | number;
  opacity?: number;
  resizeMode?: string;
  shadowColor?: string;
  shadowOffset?: WidthHeight;
  textAlignVertical?: string;
  textDecoration?: string;
  textDecorationLine?: string;
  transform?: string | TransformInstruction | Array<TransformInstruction>;
  transformMatrix?: Array<number>;
  verticalAlign?: string;
};

const emptyObject = {};

const supportsCSS3TextDecoration =
  window.CSS != null &&
  window.CSS.supports != null &&
  (window.CSS.supports('text-decoration-line', 'none') ||
    window.CSS.supports('-webkit-text-decoration-line', 'none'));

/**
 * Transform
 */

// { scale: 2 } => 'scale(2)'
// { translateX: 20 } => 'translateX(20px)'
const mapTransform = (transform: TransformInstruction) => {
  const type = Object.keys(transform)[0];
  const value = normalizeValueWithProperty(transform[type], type);
  return `${type}(${value})`;
};

// [1,2,3,4,5,6] => 'matrix3d(1,2,3,4,5,6)'
const convertTransformMatrix = (transformMatrix: Array<number>) => {
  const matrix = transformMatrix.join(',');
  return `matrix3d(${matrix})`;
};

const resolveTransform = (resolvedStyle: RNStyleObj, style: RNStyleObj) => {
  let transform = style.transform;
  if (Array.isArray(style.transform)) {
    transform = style.transform.map(mapTransform).join(' ');
  } else if (style.transformMatrix) {
    transform = convertTransformMatrix(style.transformMatrix);
  }
  resolvedStyle.transform = transform;
};

/**
 * Reducer
 */

export const convertReactNativeStyle = (style: RNStyleObj) => {
  if (!style) {
    return emptyObject;
  }

  const resolvedStyle: RNStyleObj = {};

  Object.keys(style)
    .sort()
    .forEach(prop => {
      const value = normalizeValueWithProperty(style[prop], prop);

      // Ignore everything else with a null value
      if (value == null) {
        return;
      }

      const settableProp = toCSSProp(prop);

      switch (prop) {
        // Ignore some React Native styles
        case 'aspectRatio':
        case 'elevation':
        case 'overlayColor':
        case 'resizeMode':
        case 'tintColor':

        // Ignore properties handled by yoga
        case 'alignContent':
        case 'alignItems':
        case 'alignSelf':
        case 'aspectRatio':
        case 'bottom':
        case 'direction':
        case 'display':
        case 'end':
        case 'flex':
        case 'flexBasis':
        case 'flexDirection':
        case 'flexGrow':
        case 'flexShrink':
        case 'flexWrap':
        case 'height':
        case 'justifyContent':
        case 'left':
        case 'margin':
        case 'marginBottom':
        case 'marginEnd':
        case 'marginHorizontal':
        case 'marginLeft':
        case 'marginRight':
        case 'marginStart':
        case 'marginTop':
        case 'marginVertical':
        case 'maxHeight':
        case 'maxWidth':
        case 'minHeight':
        case 'minWidth':
        case 'padding':
        case 'paddingBottom':
        case 'paddingEnd':
        case 'paddingHorizontal':
        case 'paddingLeft':
        case 'paddingRight':
        case 'paddingStart':
        case 'paddingTop':
        case 'paddingVertical':
        case 'position':
        case 'right':
        case 'start':
        case 'top':
        case 'width':
        case 'zIndex': {
          break;
        }

        // Ignore the flex property since we convert it in primitives.ts
        case 'flex': {
          break;
        }

        case 'font': {
          resolvedStyle[settableProp] = value.replace('System', SYSTEM_FONT_STACK);
          break;
        }

        case 'fontFamily': {
          if (value.indexOf('System') > -1) {
            const stack = value.split(/,\s*/);
            stack[stack.indexOf('System')] = SYSTEM_FONT_STACK;
            resolvedStyle[settableProp] = stack.join(',');
          } else if (value === 'monospace') {
            resolvedStyle[settableProp] = MONOSPACE_FONT_STACK;
          } else {
            resolvedStyle[settableProp] = value;
          }
          break;
        }

        case 'fontVariant': {
          if (Array.isArray(value) && value.length > 0) {
            resolvedStyle.fontVariant = value.join(' ');
          }
          break;
        }

        case 'textAlignVertical': {
          resolvedStyle.verticalAlign = value === 'center' ? 'middle' : value;
          break;
        }

        case 'textDecorationLine': {
          // use 'text-decoration' for browsers that only support CSS2
          // text-decoration (e.g., IE, Edge)
          if (!supportsCSS3TextDecoration) {
            resolvedStyle.textDecoration = value;
          } else {
            resolvedStyle.textDecorationLine = value;
          }
          break;
        }

        case 'transform':
        case 'transformMatrix': {
          resolveTransform(resolvedStyle, style);
          break;
        }

        case 'writingDirection': {
          resolvedStyle.direction = value;
          break;
        }

        default: {
          const longFormProperties = STYLE_SHORT_FORM_EXPANSIONS[prop];
          if (longFormProperties) {
            longFormProperties.forEach((longForm, i) => {
              // The value of any longform property in the original styles takes
              // precedence over the shortform's value.
              if (typeof style[longForm] === 'undefined') {
                resolvedStyle[toCSSProp(longForm)] = value;
              }
            });
          } else {
            resolvedStyle[settableProp] = Array.isArray(value) ? value.join(',') : value;
          }
        }
      }
    });

  return resolvedStyle;
};
