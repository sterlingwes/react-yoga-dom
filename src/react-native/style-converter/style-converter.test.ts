import { convertReactNativeStyle } from './style-converter';

const reactNativeStyle = {
  boxShadow: '1px 1px 1px 1px #000',
  borderWidthLeft: 2,
  borderWidth: 1,
  borderWidthRight: 3,
  display: 'flex',
  marginVertical: 0,
  opacity: 0,
  shadowColor: 'red',
  shadowOffset: { width: 1, height: 2 },
  resizeMode: 'contain',
};

describe('style-converter/convertReactNativeStyle', () => {
  test('noop on DOM styles', () => {
    const firstStyle = convertReactNativeStyle(reactNativeStyle);
    const secondStyle = convertReactNativeStyle(firstStyle);
    expect(firstStyle).toEqual(secondStyle);
  });

  test('shortform -> longform', () => {
    const style = {
      borderStyle: 'solid',
      boxSizing: 'border-box',
      borderBottomColor: 'white',
      borderBottomWidth: 1,
      borderWidth: 0,
      marginTop: 50,
      marginVertical: 25,
      margin: 10,
      overflow: 'hidden',
      overscrollBehavior: 'contain',
    };

    expect(convertReactNativeStyle(style)).toMatchSnapshot();
  });

  describe('flexbox styles', () => {
    test('flex: -1', () => {
      expect(convertReactNativeStyle({ flex: -1 })).toEqual({});
    });

    test('flex: 0', () => {
      expect(convertReactNativeStyle({ flex: 0 })).toEqual({});
    });

    test('flex: 1', () => {
      expect(convertReactNativeStyle({ flex: 1 })).toEqual({});
    });

    test('flex: 10', () => {
      expect(convertReactNativeStyle({ flex: 10 })).toEqual({});
    });

    test('flexBasis overrides', () => {
      expect(convertReactNativeStyle({ flexBasis: '25%' })).toEqual({});
      expect(convertReactNativeStyle({ flex: 1, flexBasis: '25%' })).toEqual({});
    });

    test('flexShrink overrides', () => {
      expect(convertReactNativeStyle({ flexShrink: 1 })).toEqual({});
      expect(convertReactNativeStyle({ flex: 1, flexShrink: 2 })).toEqual({});
    });
  });

  describe('fontFamily', () => {
    test('general case', () => {
      expect(convertReactNativeStyle({ fontFamily: 'Georgia, Times, serif' })).toMatchSnapshot();
    });

    test('"monospace"', () => {
      expect(convertReactNativeStyle({ fontFamily: 'monospace' })).toMatchSnapshot();
    });

    test('"System"', () => {
      expect(convertReactNativeStyle({ fontFamily: 'System' })).toMatchSnapshot();
      expect(convertReactNativeStyle({ font: '14px System' })).toMatchSnapshot();
    });

    test('"Noto, System"', () => {
      expect(convertReactNativeStyle({ fontFamily: 'Noto, System' })).toMatchSnapshot();
      expect(convertReactNativeStyle({ font: '14px Noto, System' })).toMatchSnapshot();
    });

    test('"Noto, BlinkMacSystemFont"', () => {
      expect(convertReactNativeStyle({ fontFamily: 'Noto, BlinkMacSystemFont' })).toMatchSnapshot();
    });
  });

  test('fontVariant', () => {
    expect(convertReactNativeStyle({ fontVariant: ['common-ligatures', 'small-caps'] })).toEqual({
      fontVariant: 'common-ligatures small-caps',
    });
  });

  test('textAlignVertical', () => {
    expect(
      convertReactNativeStyle({
        textAlignVertical: 'center',
      }),
    ).toEqual({
      verticalAlign: 'middle',
    });
  });

  describe('transform', () => {
    // passthrough if transform value is ever a string
    test('string', () => {
      const transform = 'perspective(50px) scaleX(20) translateX(20px) rotate(20deg)';
      const style = { transform };
      const resolved = convertReactNativeStyle(style);

      expect(resolved).toEqual({ transform });
    });

    test('array', () => {
      const style = {
        transform: [{ perspective: 50 }, { scaleX: 20 }, { translateX: 20 }, { rotate: '20deg' }],
      };
      const resolved = convertReactNativeStyle(style);

      expect(resolved).toEqual({
        transform: 'perspective(50px) scaleX(20) translateX(20px) rotate(20deg)',
      });
    });

    test('transformMatrix', () => {
      const style = { transformMatrix: [1, 2, 3, 4, 5, 6] };
      const resolved = convertReactNativeStyle(style);

      expect(resolved).toEqual({
        transform: 'matrix3d(1,2,3,4,5,6)',
      });
    });
  });
});
