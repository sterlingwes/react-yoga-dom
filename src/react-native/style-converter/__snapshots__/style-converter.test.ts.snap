// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`style-converter/convertReactNativeStyle fontFamily "Noto, BlinkMacSystemFont" 1`] = `
Object {
  "fontFamily": "Noto,BlinkMacSystemFont",
}
`;

exports[`style-converter/convertReactNativeStyle fontFamily "Noto, System" 1`] = `
Object {
  "fontFamily": "Noto,system-ui,-apple-system,BlinkMacSystemFont,\\"Segoe UI\\",Roboto,Ubuntu,\\"Helvetica Neue\\",sans-serif",
}
`;

exports[`style-converter/convertReactNativeStyle fontFamily "Noto, System" 2`] = `
Object {
  "font": "14px Noto, system-ui,-apple-system,BlinkMacSystemFont,\\"Segoe UI\\",Roboto,Ubuntu,\\"Helvetica Neue\\",sans-serif",
}
`;

exports[`style-converter/convertReactNativeStyle fontFamily "System" 1`] = `
Object {
  "fontFamily": "system-ui,-apple-system,BlinkMacSystemFont,\\"Segoe UI\\",Roboto,Ubuntu,\\"Helvetica Neue\\",sans-serif",
}
`;

exports[`style-converter/convertReactNativeStyle fontFamily "System" 2`] = `
Object {
  "font": "14px system-ui,-apple-system,BlinkMacSystemFont,\\"Segoe UI\\",Roboto,Ubuntu,\\"Helvetica Neue\\",sans-serif",
}
`;

exports[`style-converter/convertReactNativeStyle fontFamily "monospace" 1`] = `
Object {
  "fontFamily": "monospace,monospace",
}
`;

exports[`style-converter/convertReactNativeStyle fontFamily general case 1`] = `
Object {
  "fontFamily": "Georgia, Times, serif",
}
`;

exports[`style-converter/convertReactNativeStyle shortform -> longform 1`] = `
Object {
  "borderBottomColor": "rgba(255,255,255,1.00)",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "1px",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "0px",
  "borderRightStyle": "solid",
  "borderRightWidth": "0px",
  "borderTopStyle": "solid",
  "borderTopWidth": "0px",
  "boxSizing": "border-box",
  "marginBottom": "25px",
  "marginLeft": "10px",
  "marginRight": "10px",
  "marginTop": "50px",
  "overflowX": "hidden",
  "overflowY": "hidden",
  "overscrollBehaviorX": "contain",
  "overscrollBehaviorY": "contain",
}
`;
